api/configuration/v1alpha1/kong_target_types.go

```diff
@@ -0,0 +1,151 @@
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.spec.controlPlaneRef) || has(self.spec.controlPlaneRef)", message="controlPlaneRef is required once set"
+// +kubebuilder:validation:XValidation:rule="(!self.status.conditions.exists(c, c.type == 'Programmed' && c.status == 'True')) ? true : oldSelf.spec.controlPlaneRef == self.spec.controlPlaneRef", message="spec.controlPlaneRef is immutable when an entity is already Programmed"
+// +kubebuilder:validation:XValidation:rule="oldSelf.spec.upstreamRef == self.spec.upstreamRef", message="upstreamRef is immutable"
+type KongTarget struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec KongTargetSpec `json:"spec"`
+	// +kubebuilder:default={conditions: {{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status KongTargetStatus `json:"status"`
+}
+
+func (t *KongTarget) initKonnectStatus() {
+	t.Status.Konnect = &konnectv1alpha1.KonnectEntityStatusWithControlPlaneRef{}
+}
+
+// GetKonnectStatus returns the Konnect status contained in the KongTarget status.
+func (t *KongTarget) GetKonnectStatus() *konnectv1alpha1.KonnectEntityStatus {
+	if t.Status.Konnect == nil {
+		return nil
+	}
+	return &t.Status.Konnect.KonnectEntityStatus
+}
+
+// GetKonnectID returns the Konnect ID in the KongTarget status.
+func (t *KongTarget) GetKonnectID() string {
+	if t.Status.Konnect == nil {
+		return ""
+	}
+	return t.Status.Konnect.ID
+}
+
+// SetKonnectID sets the Konnect ID in the KongTarget status.
+func (t *KongTarget) SetKonnectID(id string) {
+	if t.Status.Konnect == nil {
+		t.initKonnectStatus()
+	}
+	t.Status.Konnect.ID = id
+}
+
+// GetControlPlaneID returns the ControlPlane ID in the KongTarget status.
+func (t *KongTarget) GetControlPlaneID() string {
+	if t.Status.Konnect == nil {
+		return ""
+	}
+	return t.Status.Konnect.ControlPlaneID
+}
+
+// SetControlPlaneID sets the ControlPlane ID in the KongTarget status.
+func (t *KongTarget) SetControlPlaneID(id string) {
+	if t.Status.Konnect == nil {
+		t.initKonnectStatus()
+	}
+	t.Status.Konnect.ControlPlaneID = id
+}
+
+// GetTypeName returns the KongTarget Kind name.
+func (t KongTarget) GetTypeName() string {
+	return "KongTarget"
+}
+
+// GetConditions returns the Status Conditions.
+func (t *KongTarget) GetConditions() []metav1.Condition {
+	return t.Status.Conditions
+}
+
+// SetConditions sets the Status Conditions.
+func (t *KongTarget) SetConditions(conditions []metav1.Condition) {
+	t.Status.Conditions = conditions
+}
+
+type KongTargetSpec struct {
+	// ControlPlaneRef is a reference to a ControlPlane this KongTarget is associated with.
+	// +optional
+	ControlPlaneRef *ControlPlaneRef `json:"controlPlaneRef,omitempty"`
+	// UpstreamRef is a reference to a KongUpstream this KongTarget is attached to.
+	// +required
+	UpstreamRef *TargetRef `json:"upstreamRef"`
```

Comment:
If this is required I believe we can change the type to be non-pointer, WDYT?

Reply:
Done.